hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/ByteBufferHeapData.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright (c) 2008-2017, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.serialization.impl;
+
+import com.hazelcast.nio.serialization.Data;
+
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+
+/**
+ * A {@link Data} implementation where the content lives on the heap, and part of a byteBuffer
+ */
+public final class ByteBufferHeapData extends com.hazelcast.internal.serialization.impl.BaseHeapData {
+
+    private final ByteBuffer payload;
+    private final int length;
+
+    public ByteBufferHeapData(byte[] bytes, int offset, int length) {
+        this.length = length;
+        this.payload = ByteBuffer.wrap(bytes, offset, length).slice();
+        payload.order(ByteOrder.BIG_ENDIAN);
+        sizeCheck();
+    }
+
+    @Override
+    public int totalSize() {
+        return length;
+    }
+
+    @Override
+    protected byte[] payload() {
+        return payload.array();
+    }
+
+    @Override
+    protected int payloadOffset() {
+        return payload.arrayOffset();
+    }
+
+    @Override
+    protected int readIntB(int offset) {
+        return payload.getInt(offset);
+    }
+
+    @Override
+    public byte[] toByteArray() {
```

Comment:
This method is not used in hot path. And in the places it is used, it is forcing me to copy the byte array. I can't return the internal byte array directly here, because it is byte array that contains the expected byte[] not itself. Only way to avoid copy here is to return ByteBuffer here(something with offset)